server/server/src/billing/stripe_client.rs

```diff
@@ -0,0 +1,437 @@
+                    }
+                }
+            _ => Err(Other(format!("Unexpected subscription status (considering payment method has already been checked): {:?}", resp)))
+        },
+        StripeResult::Err(e) => Err(match_stripe_error(&e.error).await)
+    }
+}
+
+pub async fn detach_payment_method_from_customer(
+    server_state: &ServerState,
+    payment_method_id: &str,
+) -> Result<(), StripeClientError> {
+    match send_stripe_request::<BasicStripeResponse>(
+        server_state,
+        format!(
+            "{}/{}{}",
+            STRIPE_ENDPOINT, payment_method_id, DETACH_ENDPOINT
+        ),
+        Method::POST,
+        None,
+    )
+    .await?
+    {
+        StripeResult::Ok(_) => Ok(()),
+        StripeResult::Err(e) => Err(Other(format!(
+            "Stripe returned an error whilst detaching a payment method from a customer: {:?}",
+            e
+        ))),
+    }
+}
+
+pub async fn delete_subscription(
+    server_state: &ServerState,
+    subscription_id: &str,
+) -> Result<(), StripeClientError> {
+    match send_stripe_request::<BasicStripeResponse>(
+        server_state,
+        format!(
+            "{}{}/{}",
+            STRIPE_ENDPOINT, SUBSCRIPTIONS_ENDPOINT, subscription_id
+        ),
+        Method::DELETE,
+        None,
+    )
+    .await?
+    {
+        StripeResult::Ok(_) => Ok(()),
+        StripeResult::Err(e) => Err(Other(format!(
+            "Stripe returned an error whilst deleting a subscription: {:?}",
+            e
+        ))),
+    }
+}
+
+pub async fn retrieve_invoice(
+    server_state: &ServerState,
+    invoice_id: &str,
+) -> Result<StripeInvoice, StripeClientError> {
+    let mut retrieve_subscription = HashMap::new();
+    retrieve_subscription.insert("expand[]", "subscription");
+
+    match send_stripe_request::<StripeInvoice>(
+        server_state,
+        format!("{}{}/{}", STRIPE_ENDPOINT, INVOICES_ENDPOINT, invoice_id),
+        Method::GET,
+        Some(retrieve_subscription),
+    )
+    .await?
+    {
+        StripeResult::Ok(resp) => Ok(resp),
+        StripeResult::Err(e) => Err(match_stripe_error(&e.error).await),
+    }
+}
+
+async fn send_stripe_request<U: DeserializeOwned>(
+    server_state: &ServerState,
+    url: String,
+    method: Method,
+    maybe_form: Option<HashMap<&str, &str>>,
+) -> Result<StripeResult<U>, StripeClientError> {
+    let request = server_state
+        .stripe_client
+        .request(method, &url)
+        .basic_auth::<&str, &str>(&server_state.config.stripe.stripe_secret, None);
+
+    if let Some(form) = maybe_form {
+        request.form(&form)
+    } else {
+        request
+    }
+    .send()
+    .await
+    .map_err(|e| {
+        Other(format!(
+            "Cannot make stripe request at '{}' due to reqwest error: {:?}",
+            url, e
+        ))
+    })?
+    .json::<StripeResult<U>>()
+    .await
```

Comment:
why are there 2 awaits?

Reply:
Not totally sure, to parse it to json or retrieve the bytes request returns a future.